### PR TITLE
Communication reordering pass

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102895
* #102809
* __->__ #100762

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @soumith @desertfire @anijain2305

### CODE REVIEW

**Review by anijain2305 — APPROVED** [review]

**Comment by Chillee** [comment]
Somehow merged https://github.com/pytorch/pytorch/pull/100778 into this. Oh well

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -275,12 +275,6 @@ def setup_compile_debug():
     compile_debug = os.environ.get("TORCH_COMPILE_DEBUG", "0") == "1"
 
     if compile_debug:
-        torch._logging.set_logs(
```
Comment (wconstab) [inline torch/_dynamo/utils.py:278]: did you mean to check these in under this PR?

File: torch/_inductor/analysis.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    return graph
+
+
+def get_runtime_snode(snode: "BaseSchedulerNode"):
+    """
+    Gets the runtime of Scheduler node. Currently somewhat of a placeholder, todo to be replaced by more sophisticated approaches.
+    """
+    if isinstance(snode.node, ir.AllReduce):
+        return 5
+    if isinstance(snode.node, ir.CollectiveKernel):
```
Comment (wconstab) [inline torch/_inductor/analysis.py:108]: Hmm, why do you handle 'CollectiveKernel' and 'AllReduce' instead just 'CollectiveKernel' or enumerating all 3 types of collective?

File: torch/_inductor/comms.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                if inp not in descendants:
+                    descendants.add(inp)
+                    new_nodes.append(inp)
+        cur_nodes = new_nodes
+    return descendants
+
+
+def decide_global_ordering_comms(nodes: List["scheduler.BaseSchedulerNode"]):
+    """
+    Just enforces the ordering that's in the input graph.
```
Comment (wconstab) [inline torch/_inductor/comms.py:90]: are we able to assert today that we maintain a stable ordering from trace-time to inductor time (through all our various passes)?

File: torch/_inductor/analysis.py (hunk 2 above)
Comment (Chillee) [reply torch/_inductor/analysis.py:108]: This is just a super fake cost model haha. For today's xlformer the allreduce is a lot faster than the allgathers. In the long run we want a better cost model for both the computation and the communication.

File: torch/_inductor/comms.py (hunk 3 above)
Comment (Chillee) [reply torch/_inductor/comms.py:90]: No, today, there is a separate inductor reordering pass that changes the ordering of the collectives. But anyways, we'll want to enforce our own ordering of the collectives with a pass here anyways.

File: torch/_inductor/lowering.py [hunk 4]
```diff
@@ -4168,6 +4168,8 @@ def _realize(x):
 
     @register_lowering(c10d_functional.wait_tensor)
     def wait(input):
+        from .utils import is_local
```
Comment (jansel) [inline torch/_inductor/lowering.py:4171]: unused?

File: torch/_inductor/utils.py [hunk 5]
```diff
@@ -680,6 +680,15 @@ def patched_compile_to_module(self):
     return result, source_codes
 
 
+def print2(*args):
```
Comment (jansel) [inline torch/_inductor/utils.py:683]: Why is this called print2?

File: torch/_inductor/utils.py [hunk 6]
```diff
@@ -680,6 +680,15 @@ def patched_compile_to_module(self):
...
+    if os.environ.get("LOCAL_RANK", -1) == "0":
+        print(*args)
+
+
+def is_local():
```
Comment (jansel) [inline torch/_inductor/utils.py:688]: unused?

File: torch/_inductor/ir.py [hunk 7]
```diff
@@ -4681,17 +4681,11 @@ def create(cls, x: "TensorBox", tag: str, ranks: List[int], group_size: int):
         )
 
     def codegen_collective(self, wrapper, output_name, input_names):
-        wrapper.writeline(
```
Comment (jansel) [inline torch/_inductor/ir.py:4684]: Are these changes intentional?

File: torch/_inductor/graph.py [hunk 8]
```diff
@@ -641,6 +641,20 @@ def codegen(self):
 
         self.scheduler = Scheduler(self.buffers)
         assert self.scheduler is not None  # mypy can't figure this out
+
+        from .analysis import create_fx_from_snodes
+
+        def reorder_nodes(nodes, ordering_pass):
```
Comment (jansel) [inline torch/_inductor/graph.py:647]: Move this to comms.py?

File: torch/_inductor/debug.py [hunk 9]
```diff
@@ -70,7 +62,8 @@ def draw_buffers(nodes, print_graph=False, fname=None):
         if isinstance(node, ir.ComputedBuffer):
             dtype = node.data.dtype
 
-        metadata = TensorMetadata(group, dtype, None, None, None, None, None)
+        requires_grad = node.meta["fusion_meta"].snode.node.__class__.__name__
```
Comment (jansel) [inline torch/_inductor/debug.py:65]: variable name doesn't seem right based on contents.

File: torch/_inductor/config.py [hunk 10]
```diff
@@ -51,6 +51,8 @@
 # enable reordering pass
 reordering = True
 
+comm_reordering = False
```
Comment (jansel) [inline torch/_inductor/config.py:54]: Add comment

File: torch/_inductor/config.py [hunk 11]
```diff
@@ -31,7 +31,7 @@
 inplace_buffers = True
 
 # allow reusing buffers for more efficient memory use
-allow_buffer_reuse = True
+allow_buffer_reuse = False
```
Comment (jansel) [inline torch/_inductor/config.py:34]: Intentional?

File: torch/_inductor/analysis.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
+import collections
+from typing import Any, List
+
+import torch
+import torch.fx as fx
+from . import ir
+
+
+def create_fx_from_snodes(snodes: List["scheduler.BaseSchedulerNode"]) -> fx.Graph:
```
Comment (jansel) [inline torch/_inductor/analysis.py:9]: Is this just a move?  Or was something changed?

File: torch/_inductor/comms.py [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+import torch.fx as fx
+from . import ir, scheduler
+from .analysis import get_runtime_snode
+from .dependencies import WeakDep
+from .utils import is_local, print2
+
+
+# Used to ensure that iterating over a set is deterministic
+def tuple_sorted(x):
+    return sorted(tuple(x), key=lambda x: x.name)
```
Comment (jansel) [inline torch/_inductor/comms.py:13]: Why is tuple() needed here:
```
>>> sorted({1, 5, 3})
[1, 3, 5]
>>> sorted(tuple({1, 5, 3}))
[1, 3, 5]
```

File: torch/_inductor/comms.py [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+    maximize overlap.
+    """
+    nodes = [node for node in nodes if "fusion_meta" in node.meta]
+    nodes = sink_waits(nodes)
+    nodes = raise_comms(nodes)
+    return [node.meta["fusion_meta"].snode for node in nodes]
+
+
+def debug_print(s=""):
+    import os
```
Comment (jansel) [inline torch/_inductor/comms.py:110]: imports to top

File: torch/_inductor/comms.py [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+        else:
+            for wait in tuple_sorted(cur_waits):
+                if node in wait.users:
+                    new_result.append(wait)
+                    cur_waits.remove(wait)
+
+            new_result.append(node)
+    for node in tuple_sorted(cur_waits):
+        new_result.append(node)
+    return new_result
```
Comment (jansel) [inline torch/_inductor/comms.py:35]: Is converting to FX actually doing anything here?  You aren't actually modifying the FX graph (the FX linked list will contain the old order).

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/utils.py (hunk 5 above)
Comment (Chillee) [reply torch/_inductor/utils.py:683]: I found it useful to have a `print` utility that only printed out on the first rank for debugging purposes. Will probably keep in the PR but rename.

File: torch/_inductor/ir.py (hunk 7 above)
Comment (Chillee) [reply torch/_inductor/ir.py:4684]: Yes, these were bugs in the previous implementation that were calling two allgathers for every allgather node :)

File: torch/_inductor/analysis.py (hunk 12 above)
Comment (Chillee) [reply torch/_inductor/analysis.py:9]: Just a move with some minor bug fixes. I moved it into here since it changed from being a debug utility to something that's load bearing in optimizations.

File: torch/_inductor/comms.py (hunk 15 above)
Comment (Chillee) [reply torch/_inductor/comms.py:35]: Converting it from SchedulerNodes to a FX graph is just used to make it easier to work with the graph. We then just return some topological sort of the FX nodes as a list of nodes.

File: torch/_inductor/config.py (hunk 11 above)
Comment (Chillee) [reply torch/_inductor/config.py:34]: Buffer reuse does need to be disabled for comms to work properly, but no, didn't mean to change the default.

File: torch/_inductor/comms.py (hunk 15 above)
Comment (jansel) [reply torch/_inductor/comms.py:35]: IMO this just adds complexity by having two duplicate graph representations.  If there are gaps in the existing graph we should improve it rather than duplicating.

Comment (Chillee) [reply torch/_inductor/comms.py:35]: Well, the existing graph isn't really a "graph" representation - there's no way of easily querying users or dependencies.

I think there's also just value in consistently working with the same graph representation between people writing passes on Aten FX graphs and people writing passes at the scheduler level.

Comment (jansel) [reply torch/_inductor/comms.py:35]: FX graph is missing typed edges between nodes, so can't represent scheduler graphs.  Let's just add `.users` and `.dependencies` to the existing graph, so the names are the same and reduce learning curve.

File: torch/_dynamo/utils.py (hunk 1 above)
Comment (Chillee) [reply torch/_dynamo/utils.py:278]: yeah
